### PR TITLE
Parameter-renaming decorator

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

xref https://github.com/matplotlib/matplotlib/pull/13117#discussion_r245767850, https://github.com/matplotlib/matplotlib/issues/11070, https://github.com/matplotlib/matplotlib/pull/12563, https://github.com/matplotlib/matplotlib/pull/12383.

attn @tacaswell @timhoffm I guess.

Edit: hum, probably needs a bit more thought to handle the case of **kwargs silently grabbing all unknown kwargs...
Edit: fixed.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
Actually @efiring argued in https://github.com/matplotlib/matplotlib/pull/12383#issuecomment-426872553 against the s->text change; I personally don't really mind either way and can revert it here if that's the agreement.
Note, though, that right now it's `plt.text(x, y, s)` but `Text(x, y, text)`; they should be made consistent.

**Comment by tacaswell** [comment]
Looks like it interacts with pyplot in a bit of a weird way too...

**Comment by efiring** [comment]
Regarding my earlier comment in favor of "s": I'm willing to back off on that.  It probably does make more sense at this point for `annotate` to match the `Text` API and use 'text'.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -6903,7 +6905,7 @@ def hist2d(self, x, y, bins=10, range=None, normed=False, weights=None,
         """
 
         h, xedges, yedges = np.histogram2d(x, y, bins=bins, range=range,
-                                           normed=normed, weights=weights)
+                                           normed=density, weights=weights)
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py]: Wouldn't it be "density=density"?

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: no, it's called normed on hist2d side until np 1.15... https://github.com/matplotlib/matplotlib/issues/11070#issuecomment-422411546

**Comment by anntzer** [comment]
I see your point @tacaswell re: pyplot.  I think everything can be made to work, just needs a bit more care...

**Comment by anntzer** [comment]
Went for a different, and in fact much simpler implementation (turns out the pyplot wrappers really constrain the design space here).

File: lib/matplotlib/cbook/deprecation.py [hunk 2]
```diff
@@ -258,3 +258,51 @@ def wrapper(*args, **kwargs):
...
+        def func(good_name): ...
+    """
+
+    if func is None:
+        return functools.partial(_rename_parameter, since, old, new)
+
+    signature = inspect.signature(func)
+    assert old not in signature.parameters, (
+        f"Matplotlib internal error: {old!r} cannot be a parameter for "
+        "{func.__name__}()")
```
Comment (QuLogic) [inline lib/matplotlib/cbook/deprecation.py]: There should be an `f` here, no?

File: lib/matplotlib/cbook/deprecation.py [hunk 3]
```diff
@@ -258,3 +258,51 @@ def wrapper(*args, **kwargs):
...
+    assert new in signature.parameters, (
+        f"Matplotlib internal error: {new!r} must be a parameter for "
+        "{func.__name__}()")
```
Comment (QuLogic) [inline lib/matplotlib/cbook/deprecation.py]: And here?

File: lib/matplotlib/cbook/deprecation.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/cbook/deprecation.py]: yes, thanks, fixed.

**Comment by tacaswell** [comment]
Could you also add a `make_paramaters_after_N_kwarg_only` as a decorator?

**Comment by anntzer** [comment]
That's actually going to be harder wrt. the pyplot wrappers, I think(?), because that decorator probably really needs to be applied on the wrapper as well (as the wrappers pass everything down by keyword, so the inner function cannot check whether the argument was passed positionally to the wrapper), and I tried quite hard (both here and in #13173) to not "store additional information" on the method to tell whether the wrapper needs additional decorators.

**Comment by anntzer** [comment]
Now that I think of it, one main argument for moving to keyword-only arguments is to make it easier to rename or to delete arguments, but now that this PR and #13173 implement these features, I guess it's not that useful anymore?

**Comment by anntzer** [comment]
Good to merge? :)

**Comment by jklymak** [comment]
I'd like @tacaswell to merge.  

I think we should also decide some sort of guidelines for renaming...  Just because this is easier dosn't mean we should break old documentation and examples on webpages...

**Comment by anntzer** [comment]
I think it should be treated as an API break like most others, although with a lower bar if we can reasonably assume that the parameter was rarely if ever passed as keyword (e.g. `matplotlib.use(arg=...)` is probably rare, whereas `hist(normed=...)` is fairly common (relative to the number of uses of hist())).

**Comment by jklymak** [comment]
ping @tacaswell  - I'll merge if don't hear from you by w/e

**Pull request merged**